File: src/lib/api/skill-execution.ts
/**
 * @file Skill 执行 API 模块
 * @description 封装 Skill 执行相关的 Tauri 命令调用
 *
 * 提供以下功能：
 * - executeSkill: 执行指定的 Skill
 * - listExecutableSkills: 列出所有可执行的 Skills
 * - getSkillDetail: 获取 Skill 详情
 *
 * @module lib/api/skill-execution
 * @requirements 3.1, 4.1, 5.1
 */

import { safeInvoke } from "@/lib/dev-bridge";

// ============================================================================
// 类型定义
// ============================================================================

/**
 * 可执行 Skill 信息
 *
 * 用于 listExecutableSkills 返回的 Skill 列表项
 */
export interface ExecutableSkillInfo {
  /** Skill 名称（唯一标识） */
  name: string;
  /** 显示名称 */
  display_name: string;
  /** Skill 描述 */
  description: string;
  /** 执行模式：prompt, workflow, agent */
  execution_mode: "prompt" | "workflow" | "agent";
  /** 是否有 workflow 定义 */
  has_workflow: boolean;
  /** 指定的 Provider（可选） */
  provider?: string;
  /** 指定的 Model（可选） */
  model?: string;
  /** 参数提示（可选） */
  argument_hint?: string;
}

/**
 * Workflow 步骤信息
 *
 * 描述 Workflow 中的单个步骤
 */
export interface WorkflowStepInfo {
  /** 步骤 ID */
  id: string;
  /** 步骤名称 */
  name: string;
  /** 依赖的步骤 ID 列表 */
  dependencies: string[];
}

/**
 * Skill 详情信息
 *
 * 包含 Skill 的完整信息，用于 getSkillDetail 返回
 */
export interface SkillDetailInfo extends ExecutableSkillInfo {
  /** Markdown 内容（System Prompt） */
  markdown_content: string;
  /** Workflow 步骤（如果有） */
  workflow_steps?: WorkflowStepInfo[];
  /** 允许的工具列表（可选） */
  allowed_tools?: string[];
  /** 使用场景说明（可选） */
  when_to_use?: string;
}

/**
 * 步骤执行结果
 *
 * 描述单个步骤的执行结果
 */
export interface StepResult {
  /** 步骤 ID */
  step_id: string;
  /** 步骤名称 */
  step_name: string;
  /** 是否成功 */
  success: boolean;
  /** 输出内容 */
  output?: string;
  /** 错误信息 */
  error?: string;
}

/**
 * Skill 执行结果
 *
 * 用于 executeSkill 返回的执行结果
 */
export interface SkillExecutionResult {
  /** 是否成功 */
  success: boolean;
  /** 最终输出 */
  output?: string;
  /** 错误信息 */
  error?: string;
  /** 已完成的步骤结果 */
  steps_completed: StepResult[];
}

/**
 * Skill 执行请求参数
 *
 * 统一对象参数，避免位置参数导致的调用混乱。
 */
export interface ExecuteSkillRequest {
  /** Skill 名称 */
  skillName: string;
  /** 用户输入 */
  userInput: string;
  /** 图片输入（可选） */
  images?: SkillExecutionImageInput[];
  /** 结构化请求上下文（可选） */
  requestContext?: Record<string, unknown>;
  /** Provider 覆盖 */
  providerOverride?: string;
  /** 模型覆盖 */
  modelOverride?: string;
  /** 执行 ID（用于事件关联） */
  executionId?: string;
  /** 会话 ID（用于上下文延续） */
  sessionId?: string;
}

export interface SkillExecutionImageInput {
  data: string;
  mediaType: string;
}

// ============================================================================
// Tauri 事件 Payload 类型
// ============================================================================

/**
 * 步骤开始事件 Payload
 *
 * 当 Skill 执行步骤开始时发送
 * 事件名: skill:step_start
 */
export interface StepStartPayload {
  /** 执行 ID */
  execution_id: string;
  /** 步骤 ID */
  step_id: string;
  /** 步骤名称 */
  step_name: string;
  /** 当前步骤序号（从 1 开始） */
  current_step: number;
  /** 总步骤数 */
  total_steps: number;
}

/**
 * 步骤完成事件 Payload
 *
 * 当 Skill 执行步骤完成时发送
 * 事件名: skill:step_complete
 */
export interface StepCompletePayload {
  /** 执行 ID */
  execution_id: string;
  /** 步骤 ID */
  step_id: string;
  /** 输出内容 */
  output: string;
}

/**
 * 步骤错误事件 Payload
 *
 * 当 Skill 执行步骤出错时发送
 * 事件名: skill:step_error
 */
export interface StepErrorPayload {
  /** 执行 ID */
  execution_id: string;
  /** 步骤 ID */
  step_id: string;
  /** 错误信息 */
  error: string;
  /** 是否会重试 */
  will_retry: boolean;
}

/**
 * 执行完成事件 Payload
 *
 * 当 Skill 执行完成时发送
 * 事件名: skill:complete
 */
export interface ExecutionCompletePayload {
  /** 执行 ID */
  execution_id: string;
  /** 是否成功 */
  success: boolean;
  /** 输出内容（成功时） */
  output?: string;
  /** 错误信息（失败时） */
  error?: string;
}

// ============================================================================
// Tauri 事件名常量
// ============================================================================

/** Skill 执行相关的 Tauri 事件名 */
export const SKILL_EVENTS = {
  /** 步骤开始事件 */
  STEP_START: "skill:step_start",
  /** 步骤完成事件 */
  STEP_COMPLETE: "skill:step_complete",
  /** 步骤错误事件 */
  STEP_ERROR: "skill:step_error",
  /** 执行完成事件 */
  COMPLETE: "skill:complete",
} as const;

// ============================================================================
// API 函数
// ============================================================================

/**
 * Skill 执行 API
 *
 * 封装 Skill 执行相关的 Tauri 命令调用
 */
export const skillExecutionApi = {
  /**
   * 执行指定的 Skill
   *
   * @param request - 执行参数
   * @returns 执行结果
   *
   * @requirements 3.1, 3.2, 3.5
   */
  async executeSkill(
    request: ExecuteSkillRequest,
  ): Promise<SkillExecutionResult> {
    return safeInvoke(
      "execute_skill",
      request as unknown as Record<string, unknown>,
    );
  },

  /**
   * 列出所有可执行的 Skills
   *
   * 返回所有可以执行的 Skills 列表，已过滤掉 disable_model_invocation=true 的 Skills
   *
   * @returns 可执行的 Skills 列表
   *
   * @requirements 4.1, 4.2, 4.3, 4.4
   */
  async listExecutableSkills(): Promise<ExecutableSkillInfo[]> {
    return safeInvoke("list_executable_skills");
  },

  /**
   * 获取 Skill 详情
   *
   * @param skillName - Skill 名称
   * @returns Skill 详情信息
   * @throws 如果 Skill 不存在则抛出错误
   *
   * @requirements 5.1, 5.2, 5.3, 5.4
   */
  async getSkillDetail(skillName: string): Promise<SkillDetailInfo> {
    return safeInvoke("get_skill_detail", { skillName });
  },
};

File: src-tauri/crates/agent/src/skill_execution.rs
use crate::protocol::AgentEvent as RuntimeAgentEvent;
use crate::{
    artifact_protocol::{
        extend_unique_artifact_protocol_paths, push_unique_artifact_protocol_path,
    },
    protocol_projection::project_runtime_event,
    AsterAgentState, SessionConfigBuilder, WriteArtifactEventEmitter,
};
use aster::agents::SessionConfig;
use aster::conversation::message::Message;
use aster::session::TurnContextOverride;
use futures::StreamExt;
use lime_skills::{ExecutionCallback, LoadedSkillDefinition};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;

pub type SkillEventEmitter = Arc<dyn Fn(String, RuntimeAgentEvent) + Send + Sync + 'static>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StepResult {
    pub step_id: String,
    pub step_name: String,
    pub success: bool,
    pub output: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillInputImage {
    pub data: String,
    #[serde(alias = "mediaType")]
    pub media_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillExecutionResult {
    pub success: bool,
    pub output: Option<String>,
    pub error: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub artifact_paths: Vec<String>,
    pub steps_completed: Vec<StepResult>,
}

pub struct SkillWorkflowExecution<'a> {
    pub aster_state: &'a AsterAgentState,
    pub skill: &'a LoadedSkillDefinition,
    pub user_input: &'a str,
    pub images: &'a [SkillInputImage],
    pub execution_id: &'a str,
    pub session_id: &'a str,
    pub callback: &'a dyn ExecutionCallback,
    pub memory_prompt: Option<&'a str>,
    pub emitter: SkillEventEmitter,
}

pub struct SkillPromptExecution<'a> {
    pub aster_state: &'a AsterAgentState,
    pub skill: &'a LoadedSkillDefinition,
    pub user_input: &'a str,
    pub images: &'a [SkillInputImage],
    pub execution_id: &'a str,
    pub session_id: &'a str,
    pub memory_prompt: Option<&'a str>,
    pub emitter: SkillEventEmitter,
}

#[derive(Debug, Clone)]
pub enum SkillExecutionError {
    SessionInitFailed(String),
}

impl std::fmt::Display for SkillExecutionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::SessionInitFailed(message) => write!(f, "{message}"),
        }
    }
}

impl std::error::Error for SkillExecutionError {}

struct StreamedSkillReply {
    output: String,
    error: Option<String>,
    artifact_paths: Vec<String>,
}

fn emit_skill_event(emitter: &SkillEventEmitter, event_name: &str, event: RuntimeAgentEvent) {
    emitter(event_name.to_string(), event);
}

fn collect_artifact_path_from_event(target: &mut Vec<String>, event: &RuntimeAgentEvent) {
    if let RuntimeAgentEvent::ArtifactSnapshot { artifact } = event {
        push_unique_artifact_protocol_path(target, artifact.file_path.as_str());
    }
}

fn build_step_system_prompt(
    skill_markdown: &str,
    step_name: &str,
    step_number: usize,
    total_steps: usize,
    step_prompt: &str,
    memory_prompt: Option<&str>,
) -> String {
    let base_prompt = format!(
        "{skill_markdown}\n\n---\n\n## 当前步骤: {step_name} ({step_number}/{total_steps})\n\n{step_prompt}"
    );
    if let Some(memory_prompt) = memory_prompt {
        format!("{base_prompt}\n\n{memory_prompt}")
    } else {
        base_prompt
    }
}

fn build_step_input(user_input: &str, accumulated_context: &str, is_first_step: bool) -> String {
    if is_first_step {
        accumulated_context.to_string()
    } else {
        format!("原始需求：{user_input}\n\n前序步骤输出：\n{accumulated_context}")
    }
}

fn build_prompt_system_prompt(skill_markdown: &str, memory_prompt: Option<&str>) -> String {
    if let Some(memory_prompt) = memory_prompt {
        format!("{skill_markdown}\n\n{memory_prompt}")
    } else {
        skill_markdown.to_string()
    }
}

fn build_user_message(user_input: &str, images: &[SkillInputImage]) -> Message {
    let mut user_message = Message::user().with_text(user_input);
    for image in images {
        user_message = user_message.with_image(image.data.clone(), image.media_type.clone());
    }
    user_message
}

fn build_skill_turn_context(skill: &LoadedSkillDefinition) -> Option<TurnContextOverride> {
    let allowed_tools = skill
        .allowed_tools
        .as_ref()
        .filter(|tools| !tools.is_empty())?;

    let mut metadata = HashMap::new();
    metadata.insert(
        "subagent".to_string(),
        json!({
            "allowed_tools": allowed_tools,
        }),
    );

    Some(TurnContextOverride {
        metadata,
        ..TurnContextOverride::default()
    })
}

async fn stream_skill_session(
    aster_state: &AsterAgentState,
    session_id: &str,
    event_name: &str,
    session_config: SessionConfig,
    user_message: Message,
    emitter: &SkillEventEmitter,
) -> Result<StreamedSkillReply, SkillExecutionError> {
    let agent_arc = aster_state.get_agent_arc();
    let guard = agent_arc.read().await;
    let agent = guard.as_ref().ok_or_else(|| {
        SkillExecutionError::SessionInitFailed("Agent not initialized".to_string())
    })?;

    let cancel_token = aster_state.create_cancel_token(session_id).await;
    let stream_result = agent
        .reply(user_message, session_config, Some(cancel_token.clone()))
        .await;

    let mut output = String::new();
    let mut error: Option<String> = None;
    let mut artifact_paths = Vec::new();
    let mut write_artifact_emitter = WriteArtifactEventEmitter::new(session_id.to_string());

    match stream_result {
        Ok(mut stream) => {
            while let Some(event_result) = stream.next().await {
                match event_result {
                    Ok(agent_event) => {
                        let runtime_events = project_runtime_event(agent_event);
                        for mut runtime_event in runtime_events {
                            let extra_events =
                                write_artifact_emitter.process_event(&mut runtime_event);
                            for extra_event in extra_events {
                                collect_artifact_path_from_event(&mut artifact_paths, &extra_event);
                                emit_skill_event(emitter, event_name, extra_event);
                            }
                            collect_artifact_path_from_event(&mut artifact_paths, &runtime_event);
                            if let RuntimeAgentEvent::TextDelta { ref text } = runtime_event {
                                output.push_str(text);
                            }
                            emit_skill_event(emitter, event_name, runtime_event);
                        }
                    }
                    Err(stream_error) => {
                        error = Some(format!("Stream error: {stream_error}"));
                        break;
                    }
                }
            }
        }
        Err(agent_error) => {
            error = Some(format!("Agent error: {agent_error}"));
        }
    }

    aster_state.remove_cancel_token(session_id).await;

    Ok(StreamedSkillReply {
        output,
        error,
        artifact_paths,
    })
}

pub async fn execute_skill_workflow(
    request: SkillWorkflowExecution<'_>,
) -> Result<SkillExecutionResult, SkillExecutionError> {
    let SkillWorkflowExecution {
        aster_state,
        skill,
        user_input,
        images,
        execution_id,
        session_id,
        callback,
        memory_prompt,
        emitter,
    } = request;
    let steps = &skill.workflow_steps;
    let total_steps = steps.len();
    let event_name = format!("skill-exec-{execution_id}");
    let mut steps_completed = Vec::new();
    let mut artifact_paths = Vec::new();
    let mut accumulated_context = user_input.to_string();
    let mut final_output = String::new();
    let skill_turn_context = build_skill_turn_context(skill);

    tracing::info!(
        "[execute_skill_workflow] 开始 workflow 执行: steps={}, skill={}",
        total_steps,
        skill.skill_name
    );

    for (idx, step) in steps.iter().enumerate() {
        let step_num = idx + 1;
        callback.on_step_start(&step.id, &step.name, step_num, total_steps);

        tracing::info!(
            "[execute_skill_workflow] 执行步骤 {}/{}: id={}, name={}",
            step_num,
            total_steps,
            step.id,
            step.name
        );

        let step_system_prompt = build_step_system_prompt(
            &skill.markdown_content,
            &step.name,
            step_num,
            total_steps,
            &step.prompt,
            memory_prompt,
        );
        let step_session_id = format!("{session_id}-step-{}", step.id);
        let mut session_config_builder = SessionConfigBuilder::new(&step_session_id)
            .system_prompt(step_system_prompt)
            .include_context_trace(true);
        if let Some(turn_context) = skill_turn_context.clone() {
            session_config_builder = session_config_builder.turn_context(turn_context);
        }
        let session_config = session_config_builder.build();
        let step_input = build_step_input(user_input, &accumulated_context, idx == 0);
        let user_message = build_user_message(&step_input, images);

        let reply = stream_skill_session(
            aster_state,
            &step_session_id,
            &event_name,
            session_config,
            user_message,
            &emitter,
        )
        .await?;
        extend_unique_artifact_protocol_paths(&mut artifact_paths, &reply.artifact_paths);

        if let Some(error) = &reply.error {
            callback.on_step_error(&step.id, error, false);
            steps_completed.push(StepResult {
                step_id: step.id.clone(),
                step_name: step.name.clone(),
                success: false,
                output: None,
                error: Some(error.clone()),
            });

            let final_error = format!("步骤 '{}' 执行失败: {}", step.name, error);
            callback.on_complete(false, None, Some(&final_error));
            emit_skill_event(
                &emitter,
                &event_name,
                RuntimeAgentEvent::FinalDone { usage: None },
            );

            return Ok(SkillExecutionResult {
                success: false,
                output: None,
                error: Some(final_error),
                artifact_paths,
                steps_completed,
            });
        }

        callback.on_step_complete(&step.id, &reply.output);
        steps_completed.push(StepResult {
            step_id: step.id.clone(),
            step_name: step.name.clone(),
            success: true,
            output: Some(reply.output.clone()),
            error: None,
        });
        accumulated_context = reply.output.clone();
        final_output = reply.output;
    }

    callback.on_complete(true, Some(&final_output), None);
    emit_skill_event(
        &emitter,
        &event_name,
        RuntimeAgentEvent::FinalDone { usage: None },
    );

    tracing::info!(
        "[execute_skill_workflow] Workflow 执行完成: skill={}, steps_completed={}",
        skill.skill_name,
        steps_completed.len()
    );

    Ok(SkillExecutionResult {
        success: true,
        output: Some(final_output),
        error: None,
        artifact_paths,
        steps_completed,
    })
}

pub async fn execute_skill_prompt(
    request: SkillPromptExecution<'_>,
) -> Result<SkillExecutionResult, SkillExecutionError> {
    let SkillPromptExecution {
        aster_state,
        skill,
        user_input,
        images,
        execution_id,
        session_id,
        memory_prompt,
        emitter,
    } = request;
    let event_name = format!("skill-exec-{execution_id}");
    let mut session_config_builder = SessionConfigBuilder::new(session_id)
        .system_prompt(build_prompt_system_prompt(
            &skill.markdown_content,
            memory_prompt,
        ))
        .include_context_trace(true);
    if let Some(turn_context) = build_skill_turn_context(skill) {
        session_config_builder = session_config_builder.turn_context(turn_context);
    }
    let session_config = session_config_builder.build();
    let user_message = build_user_message(user_input, images);
    let reply = stream_skill_session(
        aster_state,
        session_id,
        &event_name,
        session_config,
        user_message,
        &emitter,
    )
    .await?;

    if let Some(error) = reply.error {
        return Ok(SkillExecutionResult {
            success: false,
            output: None,
            error: Some(error.clone()),
            artifact_paths: reply.artifact_paths.clone(),
            steps_completed: vec![StepResult {
                step_id: "main".to_string(),
                step_name: skill.display_name.clone(),
                success: false,
                output: None,
                error: Some(error),
            }],
        });
    }

    Ok(SkillExecutionResult {
        success: true,
        output: Some(reply.output.clone()),
        error: None,
        artifact_paths: reply.artifact_paths,
        steps_completed: vec![StepResult {
            step_id: "main".to_string(),
            step_name: skill.display_name.clone(),
            success: true,
            output: Some(reply.output),
            error: None,
        }],
    })
}

#[cfg(test)]
mod tests {
    use super::build_skill_turn_context;
    use lime_skills::LoadedSkillDefinition;
    use std::collections::HashMap;

    fn build_loaded_skill(allowed_tools: Option<Vec<&str>>) -> LoadedSkillDefinition {
        LoadedSkillDefinition {
            skill_name: "image_generate".to_string(),
            display_name: "配图".to_string(),
            description: "测试 skill".to_string(),
            markdown_content: "test".to_string(),
            license: None,
            metadata: HashMap::new(),
            allowed_tools: allowed_tools.map(|tools| {
                tools
                    .into_iter()
                    .map(|tool| tool.to_string())
                    .collect::<Vec<_>>()
            }),
            argument_hint: None,
            when_to_use: None,
            when_to_use_config: None,
            model: None,
            provider: None,
            disable_model_invocation: false,
            execution_mode: "prompt".to_string(),
            workflow_ref: None,
            workflow_steps: Vec::new(),
            standard_compliance: Default::default(),
        }
    }

    #[test]
    fn build_skill_turn_context_forwards_allowed_tools_to_subagent_scope_metadata() {
        let skill = build_loaded_skill(Some(vec![
            "lime_create_image_generation_task",
            "social_generate_cover_image",
        ]));

        let turn_context = build_skill_turn_context(&skill).expect("turn context");
        assert_eq!(
            turn_context.metadata["subagent"]["allowed_tools"],
            serde_json::json!([
                "lime_create_image_generation_task",
                "social_generate_cover_image"
            ])
        );
    }

    #[test]
    fn build_skill_turn_context_skips_empty_allowed_tools() {
        let skill = build_loaded_skill(Some(Vec::new()));
        assert!(build_skill_turn_context(&skill).is_none());

        let skill = build_loaded_skill(None);
        assert!(build_skill_turn_context(&skill).is_none());
    }
}
